### ISSUE TITLE
Improve timekeeping by defining time zones on Mars

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#941]
**Is your feature request related to a problem? Please describe.**
- Human beings are used to define the beginning of the day as morning. The early hour is when the quantity of hours is small. As it progresses, the number gets bigger. At midnight, when it transitions onto the next day, the number goes to zero again.
- Having a time reference to a local activity is crucial. Say, I used to wake up at 8 am every morning and now I'm being asked to wake up one hour earlier (at 7 am) to go to school/work. 
- However, at 7 or 8 am is a time reference useful to me only. If my friend is on a Moon base and he wants to know exactly when I wake up to give me a call, it's not meaningful enough for him to know that I wake up at 7 or 8 am. He needs to know in what time zone I'm in and do the conversion so that he can call me when I'm awake.
- In other words, we not only want to keep track of time locally but also universally. We need both so that we can figure out what time we can meet remotely. 

**Describe the rationale**
- Now we are working our problem backward. We already have a Universal Martian Time Standard by default. What we don't have is a Local Time Standard. 
- In other words, we need a Time Zone system for settlers to conveniently keep track of time locally.
- The disadvantage of a local time zone is that as settlers scatter across Mars, they will no longer use one universal time standard. Transactions and other activities can become more difficult across the continent. It encourages independence.

**Describe the solution you desire to see**
- In [#743](https://github.com/mars-sim/mars-sim/issues/743), we discussed implementing a new shift system. We also mentioned the possibility of a Time Zone system.
- As @Urwumpe pointed out 
> why not a 24 time zones on an almost 24 h Mars?
- We could in fact have a 24 time zone. This [group](https://www.aresastro.com/2016/03/ares-time-zone-system/) sort of map out what a 24-hour time zone would look like on Mars.

```
Based upon the MTC standard, the Arean time zone system consists of 24 evenly spaced time zones, 
each 15° in longitude representing 1 hour intervals spanning from MTC-11 to MTC+12.
```

**Describe alternatives you've considered**
- Leave it the way it is. Stick to the universal time currently being used. 

**Specifications  (please complete the following information):**
- pre 3.6.0 build 8256

**Comment by bevans2000** [comment]
This will need applying to the ```ShiftManager``` constructor as the Shift are offset by a time according to the rotation around the planet.

**Comment by mokun** [comment]
In this recent [commit](https://github.com/mars-sim/mars-sim/commit/0e7db42cdc6e274457ef92506d98800bb2bca677), I came across this method `getSunriseSunsetTime` in `OrbitInfo`.

It almost serves as a good way to determine the beginning of a day.

Perhaps we aren't that far from implementing time zone after all.

**Comment by mokun** [comment]
I just read this recent [news](https://spacenews.com/white-house-directs-nasa-to-develop-lunar-time-standard/) and [discussions](https://news.ycombinator.com/item?id=39916102) regarding the Lunar time and near term time zone designation.

I added a Section H in our [Timekeeping wiki](https://github.com/mars-sim/mars-sim/wiki/Timekeeping) with the content below

Over the years, scientists from both NASA and ESA expressed interest in creating a lunar clock and defining a Coordinated Lunar Time (CLT) or Lunar Time Zones.

In fact, because of the the gravitational field strength differences experienced on between the Moon and Earth, time moves quicker on the Moon relative to Earth, at around 58.7 microseconds every day. 

With the on-going effort of returning human beings to the Moon's surface since Apollo 17 in 1972, accurate lunar timekeeping will be more important than ever for navigation and communication and future commercial development. 

Thus, NASA's [LunaNet](https://www.nasa.gov/directorates/somd/space-communications-navigation-program/lunanet-interoperability-specification/) has been developed as the Interoperability Specification for communication and navigation on the Moon.

**Comment by bevans2000** [comment]
There is some logic related to ShiftManager that calculates a time offset. Thus can be reused so each Settlement has time offset

**Comment by mokun** [comment]
@muybien3433 

Currently, there's no time zone per se in mars-sim, as you know.

Settlements anywhere on the surface of Mars uses the same clock. It's simple and elegant.

However, say if settlement A is at (0 N, 0 E) and settlement B is at (0 N, 90 E), then at sol 1:500, the sky over settlement B would look different from the sky over settlement B by 250 millisols.

I wonder if you can take a look into crafting the logics for having time zones on Mars, provided that's something you are interested in.

**Comment by bevans2000** [comment]
@mokun we now hold the offset to sunrise in the Settlement so all daily activities are synchronised with the movement of the sun. Not exactly a timezone but does achieve the same effect.
What else do you think is needed? I think this is done now. Maybe we need show the time offset in the Settlement Unit Window somewhere.

**Comment by mokun** [comment]
We did compute timeToSunset in getSunset() in `OrbitInfo` :

```java
	/**
	 * Estimates when the sunset is for this location.
	 * 
	 * @param location
	 * @return
	 */
	public MarsTime getSunset(Coordinates location) {
		Coordinates sunDirection = getSunDirection();
		// Find the sun theta when it decrease over time
		double sunTheta = sunDirection.getTheta();
		// Gets the longitude angle due to the specific location
		double lon = location.getTheta();
	
		// Rotate the globe 90 degree to a location further away from the sun to find the sunset delta time
		// Future: May add TWILIGHT_RADIANS below ;
		double gapTheta = sunTheta - lon + HALF_PI; 

		if (gapTheta < 0) {
			// Gone round the planet,
			gapTheta += (2 * Math.PI);
		}

		// Convert time from radian to millisols
		double timeToSunset = gapTheta * RADIANS_TO_MILLISOLS;

		return clock.getMarsTime().addTime(timeToSunset);
	}
```

**Comment by mokun** [comment]
> this group of 24 times zones on Mars doesn't work very well as it is geared towards Earth time of 24 hours.

No worry about the Earth's 24 hours. It's because we human define one full revolution around a planetary body as 360 degs. So for every 15 degs you have one zone. 15 degs * 24 zones = 360 deg. 

But it's true that it doesn't have to be 24 time zones. It could be just 10 zones and each zone will be faster (or slower) than the previous (or next) by 100 millisols.

It's up to us. 

I'd like that when we set foot on Mars, there's a bigger chance that we'll make Mars time mimic Earth time as much as possible to avoid confusion. Not sure by the 1000 millisols system would be adopted.

**Comment by mokun** [comment]
I can see you put the MCT +_ in TabPanelWeather. It's kind of an obscure place.

Shall we show the time zone in Mars Navigator when the mouse hovers over a place in the global map ?

I want to see that in Settlement Map as well.

What's next with it ?

**Comment by bevans2000** [comment]
Yes I just placed it there to show how it was changing for different Settlements. If you can find a better home just move it

**Comment by mokun** [comment]
Can we now use the time zone to cleanly define the sunrise time or the start of the day for each Settlement ?

**Comment by bevans2000** [comment]
Sunrise no because it will be different in the same mars zone.
But yes for the start of the working day. I've retrospectively added it to classes like the Shift manager that function on a working day.

**Comment by mokun** [comment]
@bevans2000 

How does the time zone work now ?

Is each time zone given time zone code ?

Can we print that out  ?

**Comment by mokun** [comment]
I'm looking at MarsZone class now.

Just add displaying the Time Zone id and the Time Offset for each settlement as follows: 

<img width="553" height="677" alt="Image" src="https://github.com/user-attachments/assets/5331eaad-257f-4bd2-8f0e-60c0cca99245" />
